### ISSUE TITLE
Summing multiple confusion matrix does not result with same confusion matrix

### ISSUE THREAD

**Issue by onurtore** [Lightning-AI/torchmetrics#1917]
## 🐛 Bug

I was comparing two confusion matrix, and was expecting them to be same however they are not same. In  the first one I generated the confusion matrix by summing individual confusion matrix in a batch. In the second one I generated the confusion matrix directly on the batch. 

```
confmat = torchmetrics.ConfusionMatrix(task="multiclass", normalize='true', num_classes=4).cuda()
compose_confmat = confmat(x, labels)

single_confmat = np.zeros((4,4), dtype=np.float32)
for i in range(len(x)):
    single_confmat += confmat(x[i][np.newaxis, :], labels[i][np.newaxis, :]).cpu().numpy()

```

X has the shape [N,C,H,W] and labels has the shape [N,H,W].  I was wondering which one is the correct one to calculate the IoU? Because the results differs a lot.

**Comment by SkafteNicki** [comment]
Hi @onurtore,
When using `normalize="true"` you cannot sum the individual confusion matrices, the normalization should happen after adding all the individual confusion matrices. That is essentially the problem here. Therefore, you should use the approach that generate on all batch data.
As an alternative, if you want to iteratively update the metric then just call the `update` method which will internally keep a un-normalized confusion matrix in memory and only normalize it when you call `compute` method.
Here is an example:
```python
from torchmetrics.classification import ConfusionMatrix
import torch

preds = torch.randn(10, 5)
target = torch.randint(5, (10,))

metric = ConfusionMatrix(task="multiclass", num_classes=5, normalize='true')
for i in range(10):
    metric.update(preds[i].unsqueeze(0), target[i].unsqueeze(0))
confmat = metric.compute()

confmat2 = metric(preds, target)
assert torch.allclose(confmat, confmat2)
```
Hope this helps :]

**Comment by onurtore** [comment]
@SkafteNicki thanks for the response.

 If I just set normalize to 'false'  for my confusion matrices then can I sum them iteratively? I do not exactly understand the difference between doing that, and using update method.

**Comment by SkafteNicki** [comment]
Yes, setting `normalize=None` you can then sum the individual confusion matrices and they will add up to doing the calculation on everything at once:
```python
from torchmetrics.classification import ConfusionMatrix
import torch

preds = torch.randn(10, 5)
target = torch.randint(5, (10,))

confmat = torch.zeros(5, 5).long()
metric = ConfusionMatrix(task="multiclass", num_classes=5, normalize=None)
for i in range(10):
    confmat += metric(preds[i].unsqueeze(0), target[i].unsqueeze(0))

confmat2 = metric(preds, target)
assert torch.allclose(confmat, confmat2)
```
This is essentially what `update` method does. It store a `self.confmat` internally and just adds to that when new data is provided. The `compute` method then transform that confmat in the end.
